File: src/js/handlers/opaque.ts
import { finalizationRegistry } from "../gc";
import { registerHandler } from "../handler";
import { createLuaObject, interopType, objects, uniqueId, wrappers } from "../object";

const handler: Handler = {
    priority: 200,
    from(obj: unknown): unknown {
        if (typeof obj === "object" && interopType(obj) === "opaque") {
            wrappers.set((obj as object), (obj as any).id);
            finalizationRegistry.register((obj as object), (obj as any).id);
            return obj;
        }
        return undefined;
    },
    to(obj: unknown): unknown {
        if (typeof obj === "object" && interopType(obj) === "opaque") {
            return obj;
        }

        // A Javascript object, wrap it as opaque
        const id = uniqueId();
        objects[id] = obj as object;
        return createLuaObject("opaque", { id });
    }
}

registerHandler(handler);

File: src/lua/handlers/opaque.lua
local HANDLER = {}

HANDLER.Priority = 200

function HANDLER:From(obj)
    if Interop:ObjectType(obj) == "opaque" then
        local id = obj.id
        local object = Interop:GetObject(id)
        if object then
            -- Lua's opaque object, return it
            return object
        else
            -- Javascript's opaque object, notify JS if collected
            Interop:ListenForGC(obj, function()
                Interop:Collect(id)
            end)
            Interop.m_Wrappers[obj] = id
            return obj
        end
    end
end

function HANDLER:To(obj)
    if type(obj) == "table" and Interop:ObjectType(obj) == "opaque" then
        -- Javascript's opaque object, return it
        return obj
    end
    -- Lua's opaque object, create an interop object
    local id = Interop:UniqueID()
    Interop:RegisterObject(id, obj)
    return Interop:CreateObject("opaque", { id = id })
end